skbio/metadata/_interval.py

```diff
@@ -0,0 +1,334 @@
+# ----------------------------------------------------------------------------
+
+from ._intersection import IntervalTree
+
+
+class Interval():
+    '''Stores the position and metadata of an interval.
+
+    Parameters
+    ----------
+    intervals : list of tuple of ints
+        List of tuples representing start and end coordinates.
+    boundaries : list of tuple of bool
+        List of tuples, representing the openness of each interval.
+    metadata : dict
+        Dictionary of attributes storing information of the feature
+        such as `strand`, `gene_name` or `product`.
+    interval_metadata : object
+        A reference to the `IntervalMetadata` object that this
+        interval is associated to.
+    '''
+    def __init__(self, intervals=None, boundaries=None,
+                 metadata=None, interval_metadata=None):
+        iv = []
+        for interval in intervals:
+            iv.append(_polish_interval(interval))
+
+        if boundaries is not None:
+            self.boundaries = boundaries
+        else:
+            self.boundaries = []
+
+        if metadata is not None:
+            self.metadata = metadata
+        else:
+            self.metadata = {}
+
+        self._interval_metadata = interval_metadata
+
+        if intervals is not None:
+            self.intervals = sorted(iv)
+        else:
+            self.intervals = []
+
+        if interval_metadata is not None:
+            self._add_interval()
+
+    def _add_interval(self):
+        """ Add an interval object to the IntervalTree within
+            IntervalMetadata."""
+        # Add directly to the tree.  So no need for _is_stale_tree
+        for loc in self.intervals:
+            if loc is not None:
+                start, end = loc
+                self._interval_metadata._intervals.add(start, end, self)
+        self._interval_metadata._metadata.append(self)
+
+    def __getitem__(self, key):
+        return self.metadata[key]
+
+    def __setitem__(self, key, val):
+        self.metadata[key] = val
+
+    # This is required for creating unique sets of intervals
+    def __hash__(self):
+        return hash(tuple(sorted(self.metadata.items()) +
+                          self.intervals +
+                          self.boundaries))
+
+    def __lt__(self, other):
+        return self.intervals < other.intervals
+
+    def __gt__(self, other):
+        return self.intervals > other.intervals
+
+    def __le__(self, other):
+        return self.intervals <= other.intervals
+
+    def __ge__(self, other):
+        return self.intervals >= other.intervals
+
+    def __eq__(self, other):
+        return ((self.metadata == other.metadata) and
+                (self.intervals == other.intervals) and
+                (self.boundaries == other.boundaries))
+
+    def __contains__(self, key):
+        return key in self.metadata
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __repr__(self):
+        return ''.join([
```

Comment:
👍 for including `boundaries`.

For `interval_metadata` that would be tougher to produce an eval-able repr. I don't think it's strictly necessary to have an eval-able repr even if using that style (other reasonable Python developers seem okay with this, e.g. [here](http://stackoverflow.com/a/2626364/3776794)). How about indicating whether the `Interval` object is in a dropped state (or not) instead of repr-ing `interval_metadata`? That's really what `interval_metadata` represents for the `Interval` class. Dropped state info seems useful to know about for this class since certain operations are restricted when in a dropped state.